### DISCUSSION TITLE
Any Effective method to show pdf in theia ide browser mode?

### DISCUSSION THREAD

**Discussion by wss29** [eclipse-theia/theia#14562] [Q&A] [answered]
Hi,
Is there any way to open a PDF file in the browser of Theia IDE? I have tried some VS Code extensions in Theia IDE, but it seems that they only work in the browser when using localhost. If I use an IP address, they don’t work.
Also, I tried using the [PDFObject](https://pdfobject.com/api/#pdfobject-supportspdfs) library to load a PDF on my widget, follow code work
```
   const pdfContainer = document.getElementById('pdf-container');
        if (pdfContainer) {
            const url = 'https://pdfobject.com/pdf/sample-3pp.pdf';
            const options = {
                fallbackLink: 'Download PDF',
                pdfOpenParams: {
                    view: 'FitH',
                    pagemode: 'thumbs',
                    navpanes: 0,
                },
            };

            PDFObject.embed(url, pdfContainer, options);
        } else {
            console.error('PDF container not found');
        }
```
the problem is that if the URL is a pdf in my workspace on disk, for example 
```
const url = 'file:///home/ubuntu/WPS.pdf';
```
 I got an error:
“Not allowed to load local resource: file:///home/ubuntu/WPS.pdf#view=FitH&pagemode=thumbs&navpanes=0.”
It seems that the browser mode has many limitations on displaying PDFs. Does anyone have any successful experience with this?

**Comment by msujew** [comment]
Hey @wss29,

> If I use an IP address, they don’t work.

Theia webviews are built using subdomains. These don't work when using IP addresses. Please use an actual domain name for your web IDE. Alternatively, you can set the [respective environment variable](https://github.com/eclipse-theia/theia/tree/master/packages/plugin-ext#environment-variables) to `{{hostname}}`, but this leads to serious security implications, so I wouldn't recommend it.

Aside from that, you could in theory build a Theia view that uses software like `pdf.js` to render a PDF file directly. This is essentially what the VS Code extension is using as well.

**Reply by wss29** [reply]
Hi @msujew 
   I indeed try to use pdf.js, when set 
   pdfjsLib.GlobalWorkerOptions.workerSrc = new URL('pdfjs-dist/build/pdf.worker.mjs', import.meta.url).toString()
![image](https://github.com/user-attachments/assets/ad2a7b25-e997-45b4-bef2-ad5c38170515)
Is there any method in theia to know the absolute path of file after webpack?

**Reply by msujew** [reply, marked as answer]
You will need to make TypeScript ignore the error. See also:

https://github.com/eclipse-theia/theia/blob/2078cdfdec1be23a264371b3a030896b966d2d4e/packages/plugin-ext/src/hosted/browser/plugin-worker.ts#L29-L32

**Reply by wss29** [reply]
Hi @msujew  
   Thanks for the help. Before asking my last question, I had already searched for import.meta.url and found it in PluginWorker. I simply overlooked the @ts-expect-error (TS1343) comment. I let my guard down. Thanks again.
